utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +697,450 @@ impl FromStr for FixedDecimal {
+                    // of the decimal point, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+        }
+        Ok(decimal)
+    }
+    /// Internal function for parsing directly from floats using ryū
+    fn new_from_f64_raw(float: f64) -> Result<Self, Error> {
+        if !float.is_finite() {
+            return Err(Error::Limit);
+        }
+        // note: this does not heap allocate
+        let mut buf = ryu::Buffer::new();
+        let formatted = buf.format_finite(float);
+        Self::from_str(formatted)
+    }
+
+    /// Internal function to round off `n` digits
+    /// from the right
+    ///
+    /// `self` must have at least `n` digits
+    ///
+    /// This may end up adding a digit to the left!
+    ///
+    /// This will not change the number of significant digits, it simply exists
+    /// to *round* them (and will typically reduce the size of `self.digits`)
+    ///
+    /// This function is responsible for fixing `digits`, `magnitude`, and `upper_magnitude`.
+    /// It will only modify upper_magnitude when it is not large enough to fit the rounded number.
+    /// The caller may fix up `lower_magnitude` by whatever scheme it desires
+    fn round_trailing_digits(&mut self, n: u16, mode: RoundingMode) -> Result<(), Error> {
+        debug_assert!(
+            self.digits.len() >= n as usize,
+            "Attempted to round off {} digits of number that has only {}",
+            n,
+            self.digits.len()
+        );
+        if n == 0 {
+            // no point attempting to round off any digits
+            return Ok(());
+        }
+
+        let cutoff = self.digits.len() - n as usize;
+
+        match mode {
+            RoundingMode::Unnecessary => {
+                for digit in &self.digits[cutoff..] {
+                    if *digit != 0 {
+                        return Err(Error::Limit);
+                    }
+                }
+                return Ok(());
+            }
+            RoundingMode::Truncate => {
+                self.digits.truncate(cutoff);
+                return Ok(());
+            }
+            // continue to rest of routine
+            RoundingMode::HalfExpand => (),
+        }
+
+        // Do we need to round our significant digits?
+        let round = self.digits[cutoff] >= 5;
+
+        self.digits.truncate(cutoff);
+
+        if round {
+            for digit in self.digits[..cutoff].iter_mut().rev() {
+                if *digit == 9 {
+                    // We need to round the next digit
+                    *digit = 0;
+                } else {
+                    // We need to update this digit, then we're done
+                    *digit += 1;
+                    return Ok(());
+                }
+            }
+
+            // If we reached this point then the last digit was 9 and we need to insert
+            // another digit at the beginning
+
+            self.digits.insert(0, 1u8);
```

Comment:
Suggestion: the only time we hit this case is when the entire number becomes 1 followed by trailing zeros, in which case you don't need to `.insert`; you can clear out the vector and set it to `&[1]`.